symbolic-base/src/ZkFold/Base/Algebra/Polynomials/Univariate.hs

```diff
@@ -251,57 +288,143 @@ instance (Field c, Eq c) => MultiplicativeMonoid (Poly c) where
 instance (Ring c, Arbitrary c, Eq c) => Arbitrary (Poly c) where
     arbitrary = fmap toPoly $ chooseInt (0, 128) >>= \n -> V.replicateM n arbitrary
 
-lt :: Poly c -> c
-lt (P cs) = V.last cs
+---------------------------------- Fixed degree polynomials ----------------------------------
 
-deg :: Poly c -> Integer
--- | Degree of zero polynomial is `-1`
-deg (P cs) = fromIntegral (V.length cs) - 1
+newtype PolyVec c (size :: Natural) = PV (V.Vector c)
+    deriving (Eq, Show, Generic, NFData)
 
-scaleP :: Ring c => c -> Natural -> Poly c -> Poly c
-scaleP a n (P cs) = P $ V.replicate (fromIntegral n) zero V.++ fmap (a *) cs
+class
+    ( Ring c
+    , KnownNat size
+    , AdditiveGroup (pv c size)
+    ) => UnivariateRingPolyVec pv c size where
 
-qr :: (Field c, Eq c) => Poly c -> Poly c -> (Poly c, Poly c)
-qr a b = go a b zero
-    where
-        go x y q = if deg x < deg y then (q, x) else go x' y q'
-            where
-                c = lt x // lt y
-                n = fromIntegral (deg x - deg y)
-                -- ^ if `deg x < deg y`, `n` is not evaluated, so this would not error out
-                x' = x - scaleP c n y
-                q' = q + scaleP c n one
-
--- | Extended Euclidean algorithm.
-eea :: (Field c, Eq c) => Poly c -> Poly c -> (Poly c, Poly c)
-eea a b = go (a, one) (b, zero)
-    where
-        go (x, s) (y, t) = if deg y == -1 then (x, s) else go (y, t) (r, s - q * t)
-            where
-                (q, r) = qr x y
+    toPolyVec :: V.Vector c -> pv c size
 
----------------------------------- Fixed degree polynomials ----------------------------------
+    fromPolyVec :: pv c size -> V.Vector c
 
-newtype PolyVec c (size :: Natural) = PV (V.Vector c)
-    deriving (Eq, Show, Generic, NFData)
+    poly2vec :: forall poly . (UnivariateRingPolynomial poly c) => poly c -> pv c size
 
-toPolyVec :: forall c size . (Ring c, KnownNat size) => V.Vector c -> PolyVec c size
-toPolyVec = PV . V.take (fromIntegral (value @size)) . addZeros @c @size
+    vec2poly :: forall poly . (UnivariateRingPolynomial poly c) => pv c size -> poly c
 
-fromPolyVec :: PolyVec c size -> V.Vector c
-fromPolyVec (PV cs) = cs
+    -- -- p(x) = a0
+    polyVecConstant :: c -> pv c size
 
-rewrapPolyVec :: (V.Vector c -> V.Vector c) -> PolyVec c size -> PolyVec c size
-rewrapPolyVec f (PV x) = PV (f x)
+    -- p(x) = a1 * x + a0
+    polyVecLinear :: c -> c -> pv c size
+
+    -- p(x) = a2 * x^2 + a1 * x + a0
```

Comment:
In which order are `a2, a1, a0` supplied to the function? Is it like `@polyVecQuadratic a2 a1 a0 = a2 * x^2 + a1 * x + a0@` or the other way around?